src/System.Management.Automation/utils/EncodingUtils.cs

```diff
@@ -0,0 +1,136 @@
+                // no parameter passed, default to UTF8
+                return ClrFacade.GetDefaultEncoding();
+            }
+
+            Encoding foundEncoding;
+            if (encodingMap.TryGetValue(encoding, out foundEncoding))
+            {
+                return foundEncoding;
+            }
+
+            // error condition: unknown encoding value
+            string validEncodingValues = string.Join(", ", TabCompletionResults);
+            string msg = StringUtil.Format(PathUtilsStrings.OutFile_WriteToFileEncodingUnknown, encoding, validEncodingValues);
+
+            ErrorRecord errorRecord = new ErrorRecord(
+                PSTraceSource.NewArgumentException("Encoding"),
+                "WriteToFileEncodingUnknown",
+                ErrorCategory.InvalidArgument,
+                null);
+
+            errorRecord.ErrorDetails = new ErrorDetails(msg);
+            cmdlet.ThrowTerminatingError(errorRecord);
+
+            return null;
+        }
+
+    }
+
+    /// <summary>
+    /// To make it easier to specify -Encoding parameter, we add an ArgumentTransformationAttribute here.
+    /// When the input data is of type string and is valid to be converted to System.Text.Encoding, we do
+    /// the conversion and return the converted value. Otherwise, we just return the input data.
+    /// </summary>
+    internal sealed class ArgumentToEncodingTransformationAttribute : ArgumentTransformationAttribute
+    {
+        public override object Transform(EngineIntrinsics engineIntrinsics, object inputData)
+        {
+            string encodingName;
+            if (LanguagePrimitives.TryConvertTo<string>(inputData, out encodingName))
+            {
+                Encoding foundEncoding;
+                if (EncodingConversion.encodingMap.TryGetValue(encodingName, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+            return inputData;
+        }
+
+    }
+
+    internal class EncodingArgumentCompleter : IArgumentCompleter
+    {
+        /// <summary>
+        /// This allows us to actually complete the string with tab complete.
+        /// </summary>
+        public IEnumerable<CompletionResult> CompleteArgument(
+            string commandName,
+            string parameterName,
+            string wordToComplete,
+            System.Management.Automation.Language.CommandAst commandAst,
+            System.Collections.IDictionary fakeBoundParameters)
+        {
+            List<CompletionResult> encodings = new List<CompletionResult>();
+            foreach(string encoding in EncodingConversion.TabCompletionResults)
+            {
+                if (string.IsNullOrEmpty(wordToComplete))
+                {
+                    encodings.Add(new CompletionResult(encoding, encoding, CompletionResultType.Text, encoding));
+                }
+                else if (encoding.IndexOf(wordToComplete, StringComparison.InvariantCultureIgnoreCase) == 0)
```

Comment:
I think `wordToComplete` could have wildcards in it. For example, `Get-Content -Encoding *ianUnicode<tab>` ==> `Get-Content -Encoding BigEndianUnicode`

Reply:
I can add that